File: clojure/day7_part1_2015.clj

(ns some-assembly-required
  (:require [clojure.string :as str]
            [clojure.java.io :as io]))

(defn parse-int [s]
  (Long/parseUnsignedLong s))

(def max-u16 (dec (bit-shift-left 1 16)))

(defn parse-rule [line]
  (let [[rule wire] (str/split line #" -> ")]
    [wire rule]))

(defn build-graph [input]
  (into {} (map parse-rule) (str/split-lines input)))

(defn memo-dfs [graph wire memo]
  (if-let [cached (get memo wire)]
    [cached memo]
    (cond
      (re-matches #"\d+" wire)
      [(parse-int wire) memo]

      :else
      (let [rule (get graph wire)
            parts (str/split rule #" ")]
        (cond
          (= 1 (count parts))
          (let [[v m] (memo-dfs graph (first parts) memo)]
            [(bit-and v max-u16) (assoc m wire v)])

          (= "NOT" (first parts))
          (let [[v m] (memo-dfs graph (second parts) memo)
                v (bit-and (bit-xor max-u16 v) max-u16)]
            [v (assoc m wire v)])

          :else
          (let [[a op b] parts
                [va m1] (memo-dfs graph a memo)
                [vb m2] (memo-dfs graph b m1)
                v (case op
                    "AND"  (bit-and va vb)
                    "OR"   (bit-or va vb)
                    "LSHIFT" (bit-and (bit-shift-left va vb) max-u16)
                    "RSHIFT" (bit-shift-right va vb))]
            [v (assoc m2 wire v)]))))))

(defn some-assembly-required [input]
  (let [graph (build-graph input)
        [v _] (memo-dfs graph "a" {})]
    v))

(defn -main [& _]
  (println (some-assembly-required (str/trim (slurp "input.txt")))))

(-main)

File: csharp/day7_part1_2015.cs

using System;
using System.Collections.Generic;
using System.IO;

class Program
{
    static Dictionary<string, string> rules = new Dictionary<string, string>();
    static Dictionary<string, ushort> memo = new Dictionary<string, ushort>();

    static ushort Eval(string key)
    {
        if (memo.TryGetValue(key, out var cached)) return cached;

        ushort result;
        if (ushort.TryParse(key, out result))
        {
            memo[key] = result;
            return result;
        }

        var rule = rules[key];
        var parts = rule.Split(' ');

        if (parts.Length == 1)
            result = Eval(parts[0]);
        else if (parts[0] == "NOT")
            result = (ushort)~Eval(parts[1]);
        else
        {
            var left = Eval(parts[0]);
            var right = Eval(parts[2]);
            switch (parts[1])
            {
                case "AND": result = (ushort)(left & right); break;
                case "OR": result = (ushort)(left | right); break;
                case "LSHIFT": result = (ushort)(left << right); break;
                case "RSHIFT": result = (ushort)(left >> right); break;
                default: throw new Exception("Invalid op");
            }
        }

        memo[key] = result;
        return result;
    }

    static void Main()
    {
        foreach (var line in File.ReadAllLines("input.txt"))
        {
            var sides = line.Split(" -> ");
            rules[sides[1]] = sides[0];
        }
        Console.WriteLine(Eval("a"));
    }
}
